ts/webui/src/App.tsx

```diff
@@ -233,38 +179,110 @@ class App extends React.Component<{}, AppState> {
+            this.setState(() => ({ interval: 0 }));
+            this.closeTimer();
             return;
         }
 
-        this.timerId = window.setTimeout(this.refresh, this.state.interval * 1000);
+        this.startTimer();
     };
 
-    public async lastRefresh(): Promise<void> {
+    public lastRefresh = async (): Promise<void> => {
         await EXPERIMENT.update();
         await TRIALS.update(true);
         this.setState(state => ({
             experimentUpdateBroadcast: state.experimentUpdateBroadcast + 1,
             trialsUpdateBroadcast: state.trialsUpdateBroadcast + 1
         }));
-    }
+    };
+
+    public changeInterval = (interval: number): void => {
+        this.closeTimer(); // close page auto refresh
+        if (interval === 0) {
+            return;
+        }
+        this.setState({ interval }, () => {
+            this.refresh();
+        });
+    };
+
+    public changeColumn = (columnList: string[]): void => {
+        this.setState({ columnList: columnList });
+    };
+
+    public changeExpandRowIDs = (id: string, type?: string): void => {
+        const currentExpandRowIDs = this.state.expandRowIDs;
+
+        if (!currentExpandRowIDs.has(id)) {
+            currentExpandRowIDs.add(id);
+        } else {
+            if (!(type !== undefined && type === 'chart')) {
+                currentExpandRowIDs.delete(id);
+            }
+        }
+
+        this.setState({ expandRowIDs: currentExpandRowIDs });
+    };
+
+    public changeMetricGraphMode = (val: 'max' | 'min'): void => {
+        this.setState({ metricGraphMode: val });
+    };
+
+    // overview best trial module
+    public changeEntries = (entries: string): void => {
+        this.setState({ bestTrialEntries: entries });
+    };
+
+    // overview max duration unit
+    public changeMaxDurationUnit = (unit: string): void => {
+        this.setState({ maxDurationUnit: unit });
+    };
+
+    public updateOverviewPage = (): void => {
+        this.setState(state => ({
+            experimentUpdateBroadcast: state.experimentUpdateBroadcast + 1
+        }));
+    };
+
+    public updateDetailPage = async (): Promise<void> => {
+        this.setState(state => ({
+            trialsUpdateBroadcast: state.trialsUpdateBroadcast + 1
+        }));
+    };
+
+    // fetch api to update table record data
+    public refreshDetailTable = async (): Promise<void> => {
+        await TRIALS.update(true);
+        this.setState(state => ({
+            trialsUpdateBroadcast: state.trialsUpdateBroadcast + 1
+        }));
+    };
+
+    // start to refresh page automatically
+    public startTimer = (): void => {
+        this.timerId = window.setTimeout(this.refresh, this.state.interval * 1000);
+        const storeTimerList = this.state.timerIdList;
+        storeTimerList.push(this.timerId);
+        this.setState(() => ({ timerIdList: storeTimerList }));
+    };
+
+    public closeTimer = (): void => {
+        const { timerIdList } = this.state;
```

Comment:
When the timer is already closed, will closeTimer be triggered again?

Reply:
good suggestions! although the experiment status must be `RUNNING` when user kill job because we just support to kill the RUNNING job, but double check is safer

Reply:
+1 And if user set `disable refresh` by website, ui don't need start `startTimer()` after killing job.